[WIP] Decouple Analysis Polling from AnalysisCollection

alonsogarciapablo opened the pull request:
Fixes https://github.com/CartoDB/cartodb.js/issues/1788.

Remaining work:

- [ ] Do not use analysisCollection when reseting models in the analysisPoller (extract sources from layers and dataviews instead).
- [ ] No need to initialise analysisCollection in `vis#load`
- [ ] Ensure sources are marked/unmarked as referenced when creating / updating / deleting layers and dataviews.

cc: @IagoLast

alonsogarciapablo commented on src/vis/vis.js:
```diff
@@ -262,34 +259,20 @@ var VisModel = Backbone.Model.extend({
-      }
-    }
+  _getAnalysisNodeModels: function () {
+    return this._analysisCollection.models;
```
@IagoLast instead of using `this._analysisCollection` here, we need to extract sources from `this._layersCollection` and `this._dataviewsCollection`.

IagoLast commented on src/analysis/analysis-model.js:
```diff
@@ -25,6 +25,8 @@ var AnalysisModel = Model.extend({
 
     this._initBinds();
+
+    this._referencedBy = {};
```
A little comment pointing the intention would be helpful:

`// Store every layer/dataview pointing to this analysis.`

IagoLast commented on src/analysis/analysis-model.js:
```diff
@@ -68,6 +70,16 @@ var AnalysisModel = Model.extend({
     }, this);
+
+    this.bind('change:status', function () {
+      if (this._hadStatus() && this.isReady() && this.isSourceOfAnyModel()) {
```
I suppose _hadStatus checks for a previous status, can an analysis start in "ready" status?

alonsogarciapablo replied:
Analysis models don't have a `status` until they get one from Maps API.

IagoLast commented on src/analysis/analysis-model.js:
```diff
@@ -68,6 +70,16 @@ var AnalysisModel = Model.extend({
+    }, this);
+  },
+
+  _hadStatus: function () {
```
`_hasPreviousStatus` would be a better name IMO

IagoLast commented on src/analysis/analysis-poller.js:
```diff
@@ -1,22 +1,32 @@
 var BackbonePoller = require('backbone-poller');
 
-var AnalysisPoller = function () {};
+var AnalysisPoller = function () {
```
I prefer to use [function declarations](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function) instead [function expresions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/function) to keep anonymous functions away from the stack traces.

alonsogarciapablo replied:
👍

alonsogarciapablo commented:
I closed this PR by mistake. I reverted it and here's the new one: https://github.com/CartoDB/cartodb.js/pull/1807. @IagoLast addressing your comments in the new PR.
